runbot_gitlab_minimal/models/runbot_repo.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+# Copyright <2017> <Vauxoo info@vauxoo.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import os
+import logging
+import re
+import urllib
+
+import requests
+
+from openerp import models, fields
+
+_logger = logging.getLogger(__name__)
+
+
+def custom_build(func):
+    """Decorator for functions which should be overwritten only if
+    uses_gitlab is enabled in repo.
+    """
+    def custom_func(self, cr, uid, ids, *args, **kwargs):
+        custom_ids = []
+        regular_ids = []
+        ret = None
+        for repo in ids:
+            if repo.uses_gitlab:
+                custom_ids.append(repo.id)
+            else:
+                regular_ids.append(repo.id)
+            custom = self.browse(cr, uid, custom_ids)
+            regular = self.browse(cr, uid, regular_ids)
+            if custom_ids:
+                ret = func(self, cr, uid, custom, *args, **kwargs)
+            if regular_ids:
+                regular_func = getattr(super(RunbotRepo, self), func.func_name)
+                ret = regular_func(cr, uid, regular, *args, **kwargs)
+        return ret
+    return custom_func
+
+
+class RunbotRepo(models.Model):
+    _inherit = "runbot.repo"
+
+    uses_gitlab = fields.Boolean(help='Enable the ability to use gitlab '
+                                      'instead of github')
+
+    def git(self, cr, uid, ids, cmd, context=None):
+        """Rewriting the parent method to avoid deleting the merge_request the
+        gitlab"""
+        if cmd == ['fetch', '-p', 'origin', '+refs/pull/*/head:refs/pull/*']:
+            cmd.remove('-p')
+        return super(RunbotRepo, self).git(cr, uid, ids, cmd, context=context)
+
+    def update_git(self, cr, uid, repo, context=None):
+        """Download the gitlab merge request references to work with the
+        referrals of pull github"""
+        if os.path.isdir(os.path.join(repo.path, 'refs')) and repo.uses_gitlab:
+            repo.git(['fetch', '-p', 'origin',
+                      '+refs/merge-requests/*/head:refs/pull/*'])
+        return super(RunbotRepo, self).update_git(cr, uid, repo,
+                                                  context=context)
+
+    @custom_build
+    def github(self, cr, uid, ids, url, payload=None, ignore_errors=False,
+               context=None):
+        """Translate each request from github format to gitlab format
+        - Get information from merge requests
+            input: URL_GITLAB/projects/... instead of URL_GITHUB/repos/...
+            output: res['base']['ref'] = res['gitlab_base_mr']
+                    res['head']['ref'] = res['gitlab_head_mr']
+        - Report statutes
+            input: URL_GITLABL/... instead of URL_GITHUB/statuses/...
+            output: N/A
+        """
+        for repo in self.browse(cr, uid, ids, context=context):
+            is_url_merge = False
+            if not repo.token:
+                continue
+            try:
```

Comment:
Add a `# comment` about this **is a copy&paste from original method** but with translations for gitlab